torch/autograd/gradcheck.py

```diff
@@ -102,76 +114,132 @@ def get_stride(size):
             yield x_tensor, x_idx, d_idx
 
 
-def compute_gradient(fn, inputs, x, idx, delta, eps, is_mkldnn):
+def get_numerical_jacobian(fn, inputs, outputs=None, target=None, eps=1e-3,
+                           grad_out=1.0) -> List[Tuple[torch.Tensor]]:
+    """Computes the numerical jacobian for a given fn and inputs. Returns M * N jacobians
+    where M is the number of input tensors that require grad, and N is the number of output
+    float/complex tensors.
+
+    Args:
+        fn: the function to compute the jacobian for
+        inputs: inputs to `fn`
+        outputs: provide precomputed outputs to avoid one extra invocation of fn
+        target: the Tensors wrt whom Jacobians are calculated (default=`inputs`)
+        eps: the magnitude of the perturbation during finite differencing (default=`1e-3`)
+        grad_out: grad output value used to calculate gradients.
+
+    Returns:
+        M lists of N-tuples of jacobians
+
+    Note that `target` may not even be part of `input` to `fn`, so please be
+    **very careful** in this to not clone `target`.
+    """
+    jacobians: List[Tuple[torch.Tensor]] = []
+    if outputs is None:
+        outputs = _as_tuple(fn(inputs))
+    if target is None:
+        target = inputs
+    inp_indices = [i for i, a in enumerate(target) if is_tensor_like(a) and a.requires_grad]
+    for i, (inp, inp_idx) in enumerate(zip(iter_tensors(target, True), inp_indices)):
+        jacobians += [get_numerical_jacobian_helper(fn, inp, inp_idx, inputs, outputs, eps, grad_out)]
+    return jacobians
+
+
+def prepped_input(input, input_idx, entry, entry_idx):
+    # Prepares the inputs to be passed into the function while including the new modified input.
+    if input.layout == torch._mkldnn:  # type: ignore # no attr _mkldnn
+        # Convert back to mkldnn
+        if input_idx == entry_idx:
+            return entry.to_mkldnn()
+        else:
+            return input
+    elif input.layout == torch.sparse_coo:
+        # modifications to entry are reflected in input so we could've just returned `input` here
+        # but due to an issue with coalesce, we need to do an extra clone here.
+        # TODO: get rid of this extra clone once https://github.com/pytorch/pytorch/pull/52874 is landed
+        # Make this new tensor require again in case the function has hooks
+        return torch.sparse_coo_tensor(input._indices(), input._values(), input.size()).requires_grad_(True)
+    else:
+        # We cannot use entry (input.data) if we want gradgrad to work because
+        # fn (in the gradgrad case) needs to compute grad wrt input
+        return input
+
+
+def check_outputs_same_dtype_and_shape_in_neighborhood(output1, output2, idx, delta):
```

Comment:
@nikitaved Hey! Heard from @albanD that adding such a check in gradcheck might be useful (based on an issue you guys encountered recently). I was wondering if you had any more details of such a usecase, and whether what I'm adding covers it.